src/Nix/String.hs

```diff
@@ -145,13 +150,27 @@ principledMakeNixString :: Text -> S.HashSet StringContext -> NixString
 principledMakeNixString s c = NixString s c
 
 -- | A monad for accumulating string context while producing a result string.
-newtype WithStringContext a = WithStringContext (Writer (S.HashSet StringContext) a)
-  deriving (Functor, Applicative, Monad, MonadWriter (S.HashSet StringContext))
+newtype WithStringContextT m a = WithStringContextT (WriterT (S.HashSet StringContext) m a)
```

Comment:
No need to do this at this stage, but if we're going the mtl route we should consider whether we want a class here.